File: figures/08-graphs/snippet-dfs.py
def dfs(g):
    n = g.vertices()
    color = [0] * n
    discover = [0] * n
    finish = [0] * n
    parent = [-1] * n
    time = [0]

    def visit(u):
        color[u] = 1
        time[0] += 1
        discover[u] = time[0]
        for v, _ in g.edges_from(u):
            if color[v] == 0:
                parent[v] = u
                visit(v)
        color[u] = 2
        time[0] += 1
        finish[u] = time[0]

    for u in range(n):
        if color[u] == 0:
            visit(u)
    return discover, finish, parent

File: figures/08-graphs/snippet-dfs.java
    public static DFSResult dfs(Graph g) {
        int n = g.vertices();
        int[] color = new int[n], d = new int[n], f = new int[n], p = new int[n];
        Arrays.fill(p, -1);
        List<Integer> order = new ArrayList<>();
        int[] time = {0};
        for (int u = 0; u < n; u++) if (color[u] == 0) visit(g, u, color, d, f, p, time, order);
        return new DFSResult(order.stream().mapToInt(Integer::intValue).toArray(), d, f, p);
    }

    private static void visit(Graph g, int u, int[] color, int[] d, int[] f,
                              int[] p, int[] time, List<Integer> order) {
        color[u] = 1; d[u] = ++time[0]; order.add(u);
        for (Graph.Edge e : g.edgesFrom(u)) if (color[e.to()] == 0) {
            p[e.to()] = u; visit(g, e.to(), color, d, f, p, time, order);
        }
        color[u] = 2; f[u] = ++time[0];
    }
